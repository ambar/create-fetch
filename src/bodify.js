import toHeaders from './utils/toHeaders'
import toObject from './utils/toObject'

const isPlainObject = obj =>
  obj !== null &&
  typeof obj === 'object' &&
  Object.getPrototypeOf(obj) === Object.prototype

/**
 * Stringify body
 */
const bodify = () => fetch => (url, {body, ...options} = {}) => {
  let headers = toHeaders(options.headers)
  const contentType = 'content-type'
  if (body instanceof URLSearchParams) {
    // patches old browsers
    if (!headers.get(contentType)) {
      headers.set(contentType, 'application/x-www-form-urlencoded')
    }
  } else if (isPlainObject(body) || Array.isArray(body)) {
    body = JSON.stringify(body)
    if (!headers.get(contentType)) {
      headers.set(contentType, 'application/json')
    }
  }

  return fetch(url, {
    ...options,
    body,
    headers: toObject(headers),
  })
}

export default bodify
