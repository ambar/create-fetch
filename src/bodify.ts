import toHeaders from './utils/toHeaders'
import toObject from './utils/toObject'
import isPlainObject from './utils/isPlainObject'
import {FetchEnhancer} from './types'

/* TODO: override body type
type JSONValue = string | number | boolean | JSONObject | JSONArray

interface JSONObject {
  [x: string]: JSONValue
}

interface JSONArray extends Array<JSONValue> {}

type BodifyInit = {
  body: JSONObject | JSONArray | BodyInit
}
*/

/**
 * Stringify body
 */
const bodify =
  (): FetchEnhancer =>
  (fetch) =>
  (url, {body, ...options} = {}) => {
    let headers = toHeaders(options.headers)
    const contentType = 'content-type'
    if (body instanceof URLSearchParams) {
      // patches old browsers
      if (!headers.get(contentType)) {
        headers.set(
          contentType,
          'application/x-www-form-urlencoded;charset=UTF-8'
        )
      }
    } else if (isPlainObject(body) || Array.isArray(body)) {
      body = JSON.stringify(body)
      if (!headers.get(contentType)) {
        headers.set(contentType, 'application/json')
      }
    }

    return fetch(url, {
      ...options,
      body,
      headers: toObject(headers),
    })
  }

export default bodify
