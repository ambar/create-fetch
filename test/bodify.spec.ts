import bodify from '../src/bodify'

const objectSample = {username: 'Jo Jo', lang: '中文', age: undefined}
const arraySample = [
  [1, 2],
  [3, 4],
]

describe('bodify', () => {
  const fetch = jest.fn(
    (async (...args) => args) as unknown as typeof globalThis.fetch
  )

  it('should accept Headers', async () => {
    const myFetch = bodify()(fetch)
    const headers = new Headers({a: '1'})
    headers.append('a', '2')
    const result = await myFetch('/api', {body: {} as any, headers})
    expect(result).toMatchSnapshot()
    expect(result).toEqual(
      await myFetch('/api', {body: {} as any, headers: {a: '1, 2'}})
    )
    expect(result).toEqual(
      await myFetch('/api', {body: {} as any, headers: {a: '1', A: '2'}})
    )
  })

  it('should stringify json', async () => {
    const myFetch = bodify()(fetch)
    expect(await myFetch('/api', {body: objectSample as any})).toMatchSnapshot()
    expect(await myFetch('/api', {body: arraySample as any})).toMatchSnapshot()
    const null0 = Object.create(null)
    const null1 = Object.create(Object.create(null))
    const null2 = Object.create(Object.create(Object.create(null)))
    expect(await myFetch('/api', {body: null0})).toMatchSnapshot()
    expect(await myFetch('/api', {body: null1})).toMatchSnapshot()
    expect(await myFetch('/api', {body: null2})).toMatchSnapshot()
  })

  it('should stringify form', async () => {
    const myFetch = bodify()(fetch)
    expect(
      await myFetch('/api', {body: new URLSearchParams(objectSample)})
    ).toMatchSnapshot()
    expect(
      await myFetch('/api', {
        body: new URLSearchParams(arraySample as unknown as string[][]),
      })
    ).toMatchSnapshot()
  })

  it('should not stringify with non-plain-object or non-array', async () => {
    const myFetch = bodify()(fetch)
    expect(
      await myFetch('/api', {body: new (class Foo {})() as any})
    ).toMatchSnapshot()
    expect(
      await myFetch('/api', {body: new FormData() as any})
    ).toMatchSnapshot()
    expect(await myFetch('/api', {body: new Map() as any})).toMatchSnapshot()
    expect(await myFetch('/api', {body: true as any})).toMatchSnapshot()
    expect(await myFetch('/api', {body: null as any})).toMatchSnapshot()
  })

  it('should work without options', async () => {
    const myFetch = bodify()(fetch)
    expect(await myFetch('/api')).toMatchSnapshot()
  })

  it('should allow custom content-type', async () => {
    const myFetch = bodify()(fetch)
    const contentType = 'content-type'
    const formResult = await myFetch('/api', {
      body: new URLSearchParams(),
      headers: {
        [contentType]: 'application/vnd.org+form',
      },
    })
    const jsonResult = await myFetch('/api', {
      body: {} as any,
      headers: {
        [contentType]: 'application/vnd.org+json',
      },
    })
    expect(formResult).toMatchSnapshot()
    expect(jsonResult).toMatchSnapshot()
    // ignore cases
    expect(
      await myFetch('/api', {
        body: new URLSearchParams(),
        headers: {
          [contentType.toUpperCase()]: 'application/vnd.org+form',
        },
      })
    ).toEqual(formResult)
    expect(
      await myFetch('/api', {
        body: {} as any,
        headers: {
          [contentType.toUpperCase()]: 'application/vnd.org+json',
        },
      })
    ).toEqual(jsonResult)
  })
})
