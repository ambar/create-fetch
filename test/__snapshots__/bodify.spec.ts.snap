// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`bodify should accept Headers 1`] = `
[
  "/api",
  {
    "body": "{}",
    "headers": {
      "a": "1, 2",
      "content-type": "application/json",
    },
  },
]
`;

exports[`bodify should allow custom content-type 1`] = `
[
  "/api",
  {
    "body": URLSearchParams: "",
    "headers": {
      "content-type": "application/vnd.org+form",
    },
  },
]
`;

exports[`bodify should allow custom content-type 2`] = `
[
  "/api",
  {
    "body": "{}",
    "headers": {
      "content-type": "application/vnd.org+json",
    },
  },
]
`;

exports[`bodify should not stringify with non-plain-object or non-array 1`] = `
[
  "/api",
  {
    "body": Foo {},
    "headers": {},
  },
]
`;

exports[`bodify should not stringify with non-plain-object or non-array 2`] = `
[
  "/api",
  {
    "body": FormData {},
    "headers": {},
  },
]
`;

exports[`bodify should not stringify with non-plain-object or non-array 3`] = `
[
  "/api",
  {
    "body": Map {},
    "headers": {},
  },
]
`;

exports[`bodify should not stringify with non-plain-object or non-array 4`] = `
[
  "/api",
  {
    "body": true,
    "headers": {},
  },
]
`;

exports[`bodify should not stringify with non-plain-object or non-array 5`] = `
[
  "/api",
  {
    "body": null,
    "headers": {},
  },
]
`;

exports[`bodify should stringify form 1`] = `
[
  "/api",
  {
    "body": URLSearchParams: "username=Jo+Jo&lang=%E4%B8%AD%E6%96%87&age=undefined",
    "headers": {
      "content-type": "application/x-www-form-urlencoded;charset=UTF-8",
    },
  },
]
`;

exports[`bodify should stringify form 2`] = `
[
  "/api",
  {
    "body": URLSearchParams: "1=2&3=4",
    "headers": {
      "content-type": "application/x-www-form-urlencoded;charset=UTF-8",
    },
  },
]
`;

exports[`bodify should stringify json 1`] = `
[
  "/api",
  {
    "body": "{"username":"Jo Jo","lang":"中文"}",
    "headers": {
      "content-type": "application/json",
    },
  },
]
`;

exports[`bodify should stringify json 2`] = `
[
  "/api",
  {
    "body": "[[1,2],[3,4]]",
    "headers": {
      "content-type": "application/json",
    },
  },
]
`;

exports[`bodify should stringify json 3`] = `
[
  "/api",
  {
    "body": "{}",
    "headers": {
      "content-type": "application/json",
    },
  },
]
`;

exports[`bodify should stringify json 4`] = `
[
  "/api",
  {
    "body": "{}",
    "headers": {
      "content-type": "application/json",
    },
  },
]
`;

exports[`bodify should stringify json 5`] = `
[
  "/api",
  {
    "body": {},
    "headers": {},
  },
]
`;

exports[`bodify should work without options 1`] = `
[
  "/api",
  {
    "body": undefined,
    "headers": {},
  },
]
`;
